src/mame/nmk/nmk214.cpp

```diff
@@ -0,0 +1,191 @@
+		m_input_address_bitswap[7],
+		m_input_address_bitswap[6],
+		m_input_address_bitswap[5],
+		m_input_address_bitswap[4],
+		m_input_address_bitswap[3],
+		m_input_address_bitswap[2],
+		m_input_address_bitswap[1],
+		m_input_address_bitswap[0]);
+
+	// get selection address using only 3 bits of the effective address, based on the initial config selected while device initialization:
+	u8 selection_address =
+		(BIT(effective_address, selection_address_bits[m_init_config][0]) << 0)
+		| (BIT(effective_address, selection_address_bits[m_init_config][1]) << 1)
+		| (BIT(effective_address, selection_address_bits[m_init_config][2]) << 2);
+	
+	// get the bitswap selection value, using the previously computed selection address and the selected internal config values:
+	return (BIT(init_configs[m_init_config][0], selection_address) << 0)
+		| (BIT(init_configs[m_init_config][1], selection_address) << 1)
+		| (BIT(init_configs[m_init_config][2], selection_address) << 2);
+}
+
+u16 nmk214_device::decode_word(u32 addr, u16 data)
+{
+	// compute the select value to choose which bitswap apply to the input word, based on the address where the data is located
+	u8 bitswap_select = get_bitswap_select_value(addr);
+
+	return bitswap(data,
+		output_word_bitswaps[bitswap_select][15],
+		output_word_bitswaps[bitswap_select][14],
+		output_word_bitswaps[bitswap_select][13],
+		output_word_bitswaps[bitswap_select][12],
+		output_word_bitswaps[bitswap_select][11],
+		output_word_bitswaps[bitswap_select][10],
+		output_word_bitswaps[bitswap_select][9],
+		output_word_bitswaps[bitswap_select][8],
+		output_word_bitswaps[bitswap_select][7],
+		output_word_bitswaps[bitswap_select][6],
+		output_word_bitswaps[bitswap_select][5],
+		output_word_bitswaps[bitswap_select][4],
+		output_word_bitswaps[bitswap_select][3],
+		output_word_bitswaps[bitswap_select][2],
+		output_word_bitswaps[bitswap_select][1],
+		output_word_bitswaps[bitswap_select][0]);
+}
+
+u8 nmk214_device::decode_byte(u32 addr, u8 data)
+{
+	// compute the select value to choose which bitswap apply to the input byte, based on the address where the data is located
+	u8 bitswap_select = get_bitswap_select_value(addr);
+
+	return bitswap(data,
+		output_byte_bitswaps[bitswap_select][7],
+		output_byte_bitswaps[bitswap_select][6],
+		output_byte_bitswaps[bitswap_select][5],
+		output_byte_bitswaps[bitswap_select][4],
+		output_byte_bitswaps[bitswap_select][3],
+		output_byte_bitswaps[bitswap_select][2],
+		output_byte_bitswaps[bitswap_select][1],
+		output_byte_bitswaps[bitswap_select][0]);
+}
+
+void nmk214_device::set_init_config(u8 init_config)
+{
+	// store config data only if Bit 3 matches with the operation mode of the device
+	if (BIT(init_config, 3) == m_mode)
+	{
+		m_init_config = init_config & 0x7;
+		m_device_initialized = true;
+	}
+}
+
+void nmk214_device::device_start()
+{	
+	save_item(NAME(m_mode));
+	save_item(NAME(m_init_config));
+	save_item(NAME(m_device_initialized));
```

Comment:
If the mode is hard-coded, it doesn’t need to be saved.

Reply:
Removed